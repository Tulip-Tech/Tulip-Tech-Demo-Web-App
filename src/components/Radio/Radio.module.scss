@use 'src/styles/variables';
@use 'src/styles/theme';

.header {
  margin-bottom: 8px;
  font-weight: var(--body-font-weight-bold);
  text-align: left;

  > span {
    float: right;
    color: rgba(variables.$white, 0.7);
    font-weight: normal;
  }
}

.radio {
  display: flex;
  align-items: center;
  margin-bottom: 8px;

  > label {
    margin-left: calc(#{variables.$base-spacing} / 2);
    font-size: 14px;
    cursor: pointer;
  }

  > input {
    position: relative;
    width: 20px;
    height: 20px;
    margin: 0;
    border-radius: 15px;
    transition: all 0.1s;
    appearance: none;

    &:not(:checked) {
      border: 2px solid theme.$input-resting-border-color;
      &:hover {
        border-color: theme.$input-hover-border-color;
      }
    }

    &:checked {
      width: 20px;
      height: 20px;
      border: 2px solid var(--primary-color);
      &::after {
        position: absolute;
        top: 50%;
        left: 50%;
        width: 10px;
        height: 10px;
        background-color: var(--primary-color);
        border: 2px solid transparent;
        border-radius: 15px;
        transform: translateX(-50%) translateY(-50%);
        content: '';
      }
    }
  }
}
