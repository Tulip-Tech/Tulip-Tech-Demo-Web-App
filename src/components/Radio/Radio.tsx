import React from 'react';
import { useTranslation } from 'react-i18next';

import styles from './Radio.module.scss';

import HelperText from '#components/HelperText/HelperText';
import useOpaqueId from '#src/hooks/useOpaqueId';

type Props = {
  name: string;
  value?: string;
  values: string[];
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  header?: string;
  helperText?: string;
  error?: boolean;
  required?: boolean;
};

const Radio: React.FC<Props> = ({ name, onChange, header, value, values, helperText, error, required }: Props) => {
  const { t } = useTranslation('common');
  const id = useOpaqueId('radio', name);

  return (
    <div className={error ? styles.error : undefined}>
      {header ? (
        <div className={styles.header}>
          {header}
          {!required ? <span>{t('optional')}</span> : null}
        </div>
      ) : null}
      {values.map((optionValue, index) => (
        <div className={styles.radio} key={index}>
          <input value={optionValue} name={name} type="radio" id={id + index} onChange={onChange} checked={value === optionValue} required={required} />
          <label htmlFor={id + index}>{optionValue}</label>
        </div>
      ))}
      <HelperText error={error}>{helperText}</HelperText>
    </div>
  );
};

export default Radio;
