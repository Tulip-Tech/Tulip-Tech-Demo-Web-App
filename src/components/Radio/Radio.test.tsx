import React from 'react';
import { render } from '@testing-library/react';

import Radio from './Radio';

describe('<Radio>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <Radio
        name="radio"
        header={'Choose a Value'}
        onChange={vi.fn()}
        value="value1"
        values={Array.of('value1', 'value2', 'value3')}
        helperText={'This is required!'}
        error={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
