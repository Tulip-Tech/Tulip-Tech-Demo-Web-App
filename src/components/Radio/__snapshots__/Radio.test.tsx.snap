// Vitest Snapshot v1

exports[`<Radio> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="_header_d50e6e"
    >
      Choose a Value
      <span>
        optional
      </span>
    </div>
    <div
      class="_radio_d50e6e"
    >
      <input
        checked=""
        id="radio_1235_radio0"
        name="radio"
        type="radio"
        value="value1"
      />
      <label
        for="radio_1235_radio0"
      >
        value1
      </label>
    </div>
    <div
      class="_radio_d50e6e"
    >
      <input
        id="radio_1235_radio1"
        name="radio"
        type="radio"
        value="value2"
      />
      <label
        for="radio_1235_radio1"
      >
        value2
      </label>
    </div>
    <div
      class="_radio_d50e6e"
    >
      <input
        id="radio_1235_radio2"
        name="radio"
        type="radio"
        value="value3"
      />
      <label
        for="radio_1235_radio2"
      >
        value3
      </label>
    </div>
    <div
      class="_helperText_da58e0"
    >
      This is required!
    </div>
  </div>
</div>
`;
